Vienna/Sources/Main window/CustomWKWebView.swift

```diff
@@ -173,6 +178,50 @@ class CustomWKWebView: WKWebView {
         }
         return text
     }
+
+    // MARK: Text zoom
+
+    var canMakeTextLarger: Bool {
+        guard responds(to: #selector(getter: _supportsTextZoom)),
+              responds(to: #selector(setter: _textZoomFactor)),
```

Comment:
This is not correct, since `canMakeTextLarger` is using the getter of `_textZoomFactor`. Swift requires a declaration whether the setter or getter is referenced (you cannot use both within the same declaration). I think you should keep my original code here.

Reply:
But you can't make the text larger if you aren't able to change the setting. So canMakeTextLarger should test if it has read/write access, not only read access.

Reply:
Cf. my implementation of makeTextStandardSize

Reply:
My intent with `canMakeText[Larger|Smaller]` was only to check whether the text can be made larger/smaller than it currently is. I did not consider that it should also check whether the setter of `_textZoomFactor` is available.

You could resolve it like this:
```swift
    var canMakeTextLarger: Bool {
        guard responds(to: #selector(getter: _supportsTextZoom)),
              responds(to: #selector(getter: _textZoomFactor)),
              responds(to: #selector(setter: _textZoomFactor)),
              _supportsTextZoom
        else {
            return false
        }

        return Float(_textZoomFactor) < 3.0
    }
```

To me it makes more sense to keep the selector check where it is actually used, namely `‑makeText[Larger|Smaller]`. Ultimately, the selector checks are just there to prevent crashes in case the source renames/changes/removes those selectors at some point. If Apple were to rename the selector of the setter, then the user might notice this, because the text change does not work (and file a bug report), whereas if it were queried in the getter, the menu items are greyed out instead.

Reply:
The presence of this canMakeTextLarger function is currently not required. I just kept it because it was easier to do so.
But if I needed a functionality named canMakeTextLarger, I would expect it to make all required checks before returning an answer to me.
Once again, look at the makeTextStandardSize that I wrote: simple and straightforward.

Reply:
I created these properties/methods to mimic the API of [WebKit 1](https://developer.apple.com/documentation/webkit/webview):
<img width="800" alt="Screenshot 2022-05-01 at 12 45 01" src="https://user-images.githubusercontent.com/15073177/166142545-2073c7d2-460b-46f3-9389-666fd06170e9.png">



Vienna did use `canMakeText[Larger|Smaller]` in Appcontroller `‑validateMenuItem:`, but they were removed in 8a8af7a10e78584a42fb1c83555973cc5251d259, as the functionality was broken at the time.

Right now the menu items are accordingly not “greying out” when the text-size limit is reached, e.g. minimum size 0.5 or maximum size 3.0, which is what Vienna did before.

Reply:
According to the [doc](https://docs.swift.org/swift-book/ReferenceManual/Declarations.html), whenever there is a setter, there must exist a getter.
_"if you provide a setter clause, you must also provide a getter clause."_

Reply:
That documentation refers to Swift-only computed properties or subscripts, not to Objective-C property accessors called in Swift.

Unlike Swift, Objective-C allows for different getter/setter signatures, which is why the `getter:` or `setter:` prefix is needed when referring to a property signature by selector in Swift ([docs](https://developer.apple.com/documentation/swift/using_objective-c_runtime_features_in_swift)).

I do not believe that checking the setter signature implies that any getter can be “safely” called. For instance, suppose the getter is called `isEnabled` and the setter is `enabled` (which is [not uncommon in Objective-C](https://developer.apple.com/documentation/appkit/nscontrol/1428970-enabled?language=objc)). This means that in Objective-C, the getter method signature is `‑isEnabled` whereas the setter method signature is `‑setEnabled:`; but `‑enabled` is not defined. If in Swift you were trying to call the accessor `‑enabled` to retrieve the property value, the compiler would produce an error during compilation. If you were avoiding this by using a selector and calling it then that would be undefined behaviour. Swift does not shy away from throwing irrecoverable exceptions at runtime for undefined or unexpected behaviour (e.g. when accessing a nil value that is guaranteed to have a value at compilation). So if you were just checking whether the setter `‑setEnabled:` exists without checking for the presence `‑enabled`, you could still crash at runtime if you were calling the latter.

You can easily test this assumption by renaming the property to:
```objc
@interface WKWebView (Private)
@property (setter=_setTextZoomFactor:, nonatomic) double _textZoomFactor_CRASH;
@end
```

In Swift code, you still use the `_textZoomFactor_CRASH` signature to refer to either the getter or setter. At runtime, the setter `‑_setTextZoomFactor:` is still used, which is correct. However, the getter `‑_textZoomFactor_CRASH` will lead to an unhandled exception (which might cause a crash in release builds or some other bad behaviour).

Since we are importing private APIs here which are not guaranteed to have stable signatures, we cannot make any assumptions about their correctness. Hence, any call to a private setter or getter should be validated by checking whether the getter/setter selector exists.